### PR TITLE
[WIP] Refactoring code, algorithmic aspect

### PR DESCRIPTION
**Pull request by ahojukka5**
I don't find the connection between the code and articles describing it. I'm looking for a match between the papers describing the algorithm and implementation. This work is still in progress. I'm after an easy to follow and self-describing code which has a clear consistency to the articles. ``code == math``.

Clearly we have 2 or 3 different algorithms and one is recursive. Interesting!

### CODE REVIEW

**Comment by TeroFrondelius** [comment]
What happens with new tests? Can you pull commits from master to this pull request? I really like your ideology for more readable code.

**Comment by ahojukka5** [comment]
image

At least the 2d test is passing now. I think the only algorithmic change I did was how the distance is calculated in code, compare [this](https://github.com/JuliaFEM/Miniball.jl/blob/master/src/src.jl#L99) and  [this](https://github.com/JuliaFEM/Miniball.jl/blob/refactor/src/src.jl#L149).  For me it doesn't make any sense to have negative distances. And the code is not even describing what distance is calculated. I assumed it's the maximum distance from current ball center to the points.

**Comment by TeroFrondelius** [comment]
Check the 100000 unit circle test from my pull request. If those passes then the implementation is better than C++ version.

**Comment by ahojukka5** [comment]
Btw, points should be defined as transpose to get more efficient code, I think slicing `pts[:,j]` is way more efficient than `pts[j,:]`, because of how memory is reserved in Julia.

**Comment by ahojukka5** [comment]
```bash
Test Summary:                  | Pass  Total
unit circle bug failure points |    7      7
Test Summary:              |  Pass  Broken   Total
stastical unit circle test | 99650     350  100000
```

I changed `@test_broken`to `@test` to make tests of unit circle bug failure points to pass.

**Comment by ahojukka5** [comment]
With the following modification I get the following (and all tests pass):

```
@testset "stastical unit circle test" begin
    pts = Array{Float64}(5,2)
    maxtol = 0.0
    for k in 1:100000
        for i in 1:4
            angle = rand()*2*pi
            x = cos(angle)
            y = sin(angle)
            pts[i,:] = [x y]
            if i == 1
              pts[end,:] = [cos(angle + pi) sin(angle + pi)]
            end
        end

        mb = miniball(pts)
        maxtol = max(maxtol, norm(mb.center))
    end
    println("maxtol = $maxtol")
    @test maxtol < 1.0e-10
end

# output

maxtol = 6.080681775956368e-13
```

In what precision should we have origo to be (0,0)? I think the bug got resolved by refactoring the code. Now it is hopefully more human readable so finding bug gets easier. I still don't understand the content of the primitive function [here](https://github.com/JuliaFEM/Miniball.jl/blob/refactor/src/src.jl#L67). For me it doesn't look like the formulas in [1]. What I think it should do this:

image

Also the use of recursive function call looks odd to me. After reading the article it's not clear at all how this everything ends up to a recursion.

[1] https://people.inf.ethz.ch/gaertner/subdir/texts/own_work/esa99_final.pdf

File: src/src.jl [hunk 1]
```diff
@@ -10,372 +10,159 @@
...
+            v[fsize, :] -= a[fsize, i] * v[i, :]
         end
-        container.squared_radius = container.arr_squared_r[fsize]
-        container.fsize += 1
-        container.ssize = container.fsize
+
+        e_var = norm(pivot[:]-c[fsize-1, :])^2 - arr_squared_r[fsize-1]
+        z[fsize] = 2.0 * norm(v[fsize, :])^2
+        # compute z_fsize and reject push if z_fsize is too small
+        isapprox(z[fsize], 0.0) && return false
```
Comment (jw3126) [inline src/src.jl]: `isapprox(x, 0.0)` is the same as `x == 0.0` in julia. You need to add some tolerance for it to becomue useful.
e.g `isapprox(x, 0.0, atol=eps(typeof(x))` or just`abs(x) < eps(...)`.

File: src/src.jl [hunk 2]
```diff
@@ -10,372 +10,159 @@
...
+             dim = size(ball.points, 2)
+             unshift!(ball.support_vector, pivot)
+             if length(ball.support_vector) == (dim + 2)
+                 pop!(ball.support_vector)
              end
          end
+         dr2 = ball.squared_radius - r2_prev
+         println("Radius: $(sqrt(ball.squared_radius)), ch: $dr2")
+         isapprox(dr2, 0.0) && break
+         r2_prev = copy(ball.squared_radius)
```
Comment (jw3126) [inline src/src.jl]: `r2_prev = ball.squared_radius` is enough, `Float64` is immutable. This is not a `&double` or something.

**Comment by jw3126** [comment]
@ahojukka5 concerning precision of the final result, I would even be fine with `sqrt(eps(Float64))`. Very cool that the tests pass. Also the code starts looking like something that mortals can understand.

File: src/src.jl [hunk 3]
```diff
@@ -10,372 +10,159 @@
...
 
 Recursive function for creating a miniball using the support vector.
 """
- function miniball_support_points{F<:AbstractFloat}(container::MBContainer, end_support_vector::Array{F, 1})
-    dim   = container.dim
-    fsize = container.fsize
-    ssize = container.ssize
-    center = vec(container.center)
-    squared_radius = container.squared_radius
+ function miniball_support_points(container, end_support_vector)
```
Comment (jw3126) [inline src/src.jl]: I think this function mutates the container, so should end with a `!`.

File: src/src.jl (hunk 1 above)
Comment (ahojukka5) [reply src/src.jl]: In the original implementation, there is `pow_2(1e-21)`. I actually don't like this workaround at all, we should inspect in what geometric situations `f[size]` gets small and skip those situations in other parts of the code to make this stable. And assume that `f[size]` is big enough.

Comment (jw3126) [reply src/src.jl]: IIRC there was no situation, where this is smaller then `eps(Float64)` in the tests. So you could add an assertion and see what happens.

**Comment by jw3126** [comment]
Before changing https://github.com/JuliaFEM/Miniball.jl/blob/447fe2d3185c9a004fb41a2e7d293bd93edaf14f/src/src.jl#L151 to an index comparison, we should add test cases with duplicate points and see if that breaks something.

File: src/src.jl (hunk 1 above)
Comment (ahojukka5) [reply src/src.jl]: image
image

I think it's some kind of least squares approximation or similar, and when z gets zero obviously the matrix is not invertible anymore. 

"Equation (8) shows that A_B may become ill-conditioned (and the entries of A_B^-1 very large and unreliable) if z evaluates to a very small number. The subsequent lemma develops a geometric interpretation of z from which we can see that this happens exactly if the new point q_m is very close to the affine hull of the previous ones. This can be the case e.g. if input points are very close together or even equal."

There is also a solution to this problem, but it looks that it's not implemented to the code? Anyway the code is probably going to explode is set two points with same coordinates or the difference between them is `eps`.

Comment (ahojukka5) [reply src/src.jl]: At least we need a test to check that we get something reasonable when points are near each other. This is for #24.

**Comment by TeroFrondelius** [comment]
I ran the random tests myself and still few fails, but not with the corrected C++.  Looking at the diff from fixed C++, this is probably the key change: 
\-      if (max_e > nt0) {
\+      if (sqr_r < 0 || max_e > sqr_r * default_tol) {
 	// check if the pivot is already contained in the support set
 	if (std::find(L.begin(), support_end, pivot) == support_end) {
 	  assert (fsize == 0);

**Comment by TeroFrondelius** [comment]
# Deprecated
This repo has some nasty bug, see broken tests. After bug hunt a decission to refactor led to the rewriting of the package, which gave the opportunity to change the license. Development continues in  [BoundingSphere.jl](https://github.com/JuliaFEM/BoundingSphere.jl). Use that package instead.
